inferno-ml-remote/src/Inferno/ML/Remote/Types.hs

```diff
@@ -0,0 +1,216 @@
+  )
+where
+
+import Control.Applicative (Alternative ((<|>)), optional, (<**>))
+import Control.Exception (Exception (displayException), throwIO)
+import Control.Monad.Reader (ReaderT)
+import Data.Aeson
+  ( FromJSON (parseJSON),
+    ToJSON,
+    eitherDecodeFileStrict,
+    withObject,
+    (.:),
+    (.:?),
+  )
+import Data.String (IsString)
+import Data.Text (Text)
+import qualified Data.Text as Text
+import Data.Word (Word64)
+import GHC.Generics (Generic)
+import qualified Options.Applicative as Options
+import Servant (Handler, JSON, Post, ReqBody, (:>))
+
+type InfernoMlRemoteAPI =
+  "inference" :> ReqBody '[JSON] Script :> Post '[JSON] EvalResult
+
+type InfernoMlRemoteM = ReaderT InfernoMlRemoteEnv Handler
+
+newtype InfernoMlRemoteEnv = InfernoMlRemoteEnv
+  { modelCache :: Maybe ModelCacheOption
+  }
+  deriving stock (Generic)
+
+-- | Generic container for errors that may arise when parsing\/typechecking
+-- Inferno scripts in handlers. It doesn\'t matter what the specific error is
+-- as it will only be used in a 400 or 500 response with a message body containing
+-- the error
+data SomeInfernoError where
+  SomeInfernoError :: forall a. Show a => a -> SomeInfernoError
+
+deriving stock instance Show SomeInfernoError
+
+instance Exception SomeInfernoError where
+  displayException (SomeInfernoError x) = show x
+
+-- TODO
+-- Add more ways to load?
+data ModelCacheOption
+  = -- | Path to source directory holding models
+    Paths FilePath ModelCache
+  | -- | Path to source directory holding compressed models
+    CompressedPaths FilePath ModelCache
+  deriving stock (Show, Eq, Generic)
+
+instance FromJSON ModelCacheOption where
+  parseJSON = withObject "ModelCacheOption" $ \o ->
+    o .:? "compressed" >>= \case
+      Just True -> CompressedPaths <$> o .: "store" <*> o .: "cache"
+      _ -> Paths <$> o .: "store" <*> o .: "cache"
+
+-- | Options for caching ML models to be used with Inferno scripts. When a script
+-- uses @ML.loadModel@, models will be copied from the source configured in
+-- 'ModelCacheOption' and saved to the 'cache' directory. Once the `maxSize` has
+-- been exceeded, least-recently-used cached models will be removed
+data ModelCache = ModelCache
+  { -- | Directry where the models should be cached
+    path :: FilePath,
+    -- | Maximum size in bytes of the model cache directory
+    maxSize :: Word64
+  }
+  deriving stock (Show, Eq, Generic)
+
+instance FromJSON ModelCache where
+  parseJSON = withObject "ModelCache" $ \o ->
+    ModelCache <$> o .: "path" <*> o .: "max-size"
+
+data Options = Options
+  { port :: Word64,
+    modelCache :: Maybe ModelCacheOption
+  }
+  deriving stock (Show, Eq, Generic)
+
+instance FromJSON Options where
+  parseJSON = withObject "Options" $ \o ->
+    Options <$> o .: "port" <*> o .:? "model-cache"
+
+newtype Script = Script Text
+  deriving stock (Show, Generic)
+  deriving newtype (Eq, FromJSON, ToJSON, IsString)
+
+newtype EvalResult = EvalResult Text
```

Comment:
Why is the eval result `Text`? Could we use `Value` and preserve the data types of the returned result?

Reply:
There's no `ToJSON` instance. If you think we can write an instance that makes sense, I'd also prefer that

Reply:
I think we should be able to write a sensible `ToJSON`. Perhaps even the default/deriving one might be okay?

Reply:
I don't think we can derive an instance because one of the constructors holds a function (`VFun`)

Reply:
Oh yes, good catch. But that's for intermediate interpreter values only. The final result of a script will never be a function, and we check this when we type check scripts:
https://github.com/plow-technologies/inferno/blob/a143f90433bf12b3115edf4e2f7ab698c87187a2/inferno-lsp/src/Inferno/LSP/ParseInfer.hs#L538-L542

So perhaps we can serialize `VFun` to some suitable error value /  raise an exception?